sparklingpandas/dataframe.py

```diff
@@ -58,30 +58,74 @@ def fromRecords(records):
 
         return self._schema_rdd.rdd.flatMap(fromRecords)
 
-    def __evil_apply_with_dataframes(self, func):
+    def __evil_apply_with_dataframes(self, func, preservesColumns=False):
         """Convert the underlying SchmeaRDD to an RDD of DataFrames.
         apply the provide function and convert the result back.
         This is hella slow."""
         source_rdd = self._rdd()
         result_rdd = func(source_rdd)
-        return self.from_rdd_of_dataframes(result_rdd)
+        # By default we don't know what the columns & indexes are so we let
+        # from_rdd_of_dataframes look at the first partition to determine them.
+        columnsIndexes = None
+        if preservesColumns:
+            index_names = self._index_names
+            # Remove indexes from the columns
+            columns = self._schema_rdd.columns[len(self._index_names):]
+            columnsIndexes = (columns, index_names)
+        return self.from_rdd_of_dataframes(
+            result_rdd, columnsIndexes=columnsIndexes)
 
-    def from_rdd_of_dataframes(self, rdd):
-        """Take an RDD of dataframes and return a Dataframe"""
+    def _first_as_df(self):
+        """Gets the first row as a dataframe. Useful for functions like
```

Comment:
specify which type of dataframe to be extra clear. Maybe we can come up w a short hand? SDataFrame (SparkDataFrame), PDataFrame (PandasDataFrame), and DataFrame (our dataframe)?